webpack/assets/javascripts/react_app/components/bookmarks/form/index.js

```diff
@@ -0,0 +1,59 @@
+import React from 'react';
+import { reduxForm } from 'redux-form';
+import { connect } from 'react-redux';
+import Form from '../../common/forms/Form';
+import TextField from '../../common/forms/TextField';
+import { required, length } from 'redux-form-validators';
+import * as FormActions from '../../../redux/actions/common/forms';
+
+class BookmarkForm extends React.Component {
+  // if we got here, its already post client validations
+  submit({ name, query, publik }, dispatch, props) {
+    const { submitForm, url, controller } = props;
+    const values = { name, query, public: publik || false, controller };
+
+    return submitForm({url, values, item: 'Bookmark'});
+  }
+
+  render() {
+    const { handleSubmit, submitting, error, onCancel } = this.props;
+
+    return (
+      <Form
+        onSubmit={handleSubmit(this.submit)}
+        onCancel={onCancel}
+        disabled={submitting}
+        submitting={submitting}
+        error={error}
+      >
+        <TextField
+          name="name"
+          type="text"
+          required="true"
+          label={__('Name')}
+          validate={[required(), length({ max: 254 })]}
```

Comment:
Out of curiosity, shoukd we duplicate validation definitions or should we try ti generate it here based on what's defined in model? While this one is easy, my concern is with validation on uniqueness or custom validation messages. Is the plan to support it in UI too or should we leave it up to server to validate it on save?

Reply:
the way it works now is that we try to submit, if we get back a 422 error, then we use the json error message and add it to the form - see https://github.com/theforeman/foreman/pull/4807/files#diff-35521d7d9795ca3a7f62a33bb3ceef6aR20 and a test https://github.com/theforeman/foreman/pull/4807/files#diff-af7b70459125c57d18f34cd03710da9cR71

Reply:
and a demo of that:

![recorded](https://user-images.githubusercontent.com/20167/30151113-cfa73e8a-93b6-11e7-9810-fd78bb06168a.gif)

Reply:
ok that's nice. So we have two levels of validation, which I think is correct, but I'd like to have single definition if possible.